### DISCUSSION TITLE
✨ Add support for live development with running Dagger services

### DISCUSSION THREAD

**Discussion by levlaz** [dagger/dagger#12400] [Ideas]
### What are you trying to do?

I want to be able to make changes to code locally and see them reflected in my running Dagger services similar to how docker-compose and docker run `--v` works. 

In particular, the DX behind how the new docker-compose watch feature works would fit nicely in what I hope we can do with Dagger in the future. https://docs.docker.com/compose/file-watch/ 

Discussion about this can be found here: https://discord.com/channels/707636530424053791/1224583500569514024/1224583500569514024

### Why is this important to you?

The introduction of `*Service` has made it really easy to get a one liner dev environment up and running on my local laptop using the dagger cli without having to install any local dependencies. Supporting the ability to hot reload would unlock a whole new category of powerful local development use cases for all dagger users. 

### How are you currently working around this?

There is no workaround right now. :(

**Comment by sheldonhull** [comment]
Is there any alternative to this right now other than a live file watcher rerunning dagger commands? 
Being able to replace my wrapped up docker calls for live reloading of site and other tooling is critical for my further adoption. Without this most of the local service stuff that docker compose, or my wrapped up magefile style docker commands can't be replaced without losing out on the development experience.

**Comment by levlaz** [comment]
Unfortunately there is no other workaround right now. 

> Is there any alternative to this right now other than a live file watcher rerunning dagger commands? 

I had not thought of this approach, we are planning to make some
optimizations to dagger cli startup time which may make this a viable
short term option. 

> Being able to replace my wrapped up docker calls for live reloading of site and other tooling is critical for my further adoption. Without this most of the local service stuff that docker compose, or my wrapped up magefile style docker commands can't be replaced without losing out on the development experience.

I completely agree, I think this would unlock some really great local
development use cases.

**Comment by helderco** [comment]
I would try mutagen first.

**Comment by mvgijssel** [comment]
I would also love this feature directly integrated into Dagger. For now I’m going to see if I can integrate Dagger with Tilt https://tilt.dev/

**Comment by levlaz** [comment]
> I would also love this feature directly integrated into Dagger. For now I’m going to see if I can integrate Dagger with Tilt https://tilt.dev/

tilt seems interesting, I have not heard of it, please do share how it
goes if you give it a try!

**Comment by ghostsquad** [comment]
I too am considering tilt or skaffold, but would really like to not use either.

**Comment by sheldonhull** [comment]
I've used tilt but it has its own DSL (Starlark) and while having it invoke Dagger automatically could solve it short term, ideally I'd have a way to do a file watcher locally for live reloading. I think this is one of the core areas for me that I need to solve if I want to replace more of the mage/local automation workflows.

**Comment by wmcnamee-coreweave** [comment]
Just wondering if there's been any progress/movement on this front?

**Comment by bupd** [comment]
I’m setting up Dagger to run a live development server with [Air](https://github.com/air-verse/air) for a Harbor Satellite project. The goal is to have real-time file synchronization between the host and the container, similar to the way Docker handles bind mounts:

```bash
docker run -p 8080:8080 --rm -v $(pwd):/app --name my-go-docker-air my-go-docker
```

With Docker, any changes in the host files are reflected instantly inside the container. Currently, I’m using `dagger.withMountedDirectory()` to mount my local directory in Dagger, but changes in the host files aren’t synchronized with the container.

This feature would greatly improve development workflows where live reloading is essential, such as with tools like Air. Having real-time sync would streamline development in Dagger, especially for projects that depend on continuous updates from the host.

**Suggestion**: If possible, adding support for `dagger.withMountedVolume()` to mount the local directory as a true bind mount could allow Dagger to mirror Docker’s behavior, enabling real-time file syncing from host to container. This feature would be incredibly valuable for workflows that rely on live reloading, such as with tools like Air, and would greatly enhance development efficiency within Dagger.

Thanks for considering this feature request!

**Comment by shykes** [comment]
@bupd in your specific use case, do you need two-way filesync (host to container + container to host), or only one-way (host to container)?

**Comment by rawkode** [comment]
I'm not bupd, but I only need host to container 👍🏻

**Comment by bupd** [comment]
@shykes host to container is fine.

**Comment by shykes** [comment]
cc @aluzzardi @sipsma the use of `air` (go-specific watcher) is interesting. We've been discussing the idea of moving the watch logic to dagger. effectively making dagger a universal watcher. This would mean re-running the pipeline on each change to local directories. The idea is that with caching & filesync optimized enough, this approach becomes fast enough that you don't need to sacrifice the benefits of immutable inputs. And all toolchains for all languages would get a state of the art watch feature for free. Discussion still underway :)

**Comment by sagikazarmark** [comment]
Rerunning the entire pipeline is probably suboptimal when running the pipeline (ie. developing an application) involves running dependency services as well.

**Comment by shykes** [comment]
> Rerunning the entire pipeline is probably suboptimal when running the pipeline (ie. developing an application) involves running dependency services as well.

Well, it wouldn't be exactly the same as running `dagger call` twice in a row.. Because the engine itself would be "watch-aware", and perhaps could do things like keep dependency service running; as well as keeping tcp tunnels running. So for example, even though the app server would be restarted continuously, its dependencies would not, and the dagger proxy would keep all TCP connections open from the browser. So it would really boil down to cache and filesync optimization.

**Comment by beingminimal** [comment]
> cc @aluzzardi @sipsma the use of `air` (go-specific watcher) is interesting. We've been discussing the idea of moving the watch logic to dagger. effectively making dagger a universal watcher. This would mean re-running the pipeline on each change to local directories. The idea is that with caching & filesync optimized enough, this approach becomes fast enough that you don't need to sacrifice the benefits of immutable inputs. And all toolchains for all languages would get a state of the art watch feature for free. Discussion still underway :)

Fantastic! Our team has been eagerly awaiting Dagger's complete DevOps lifecycle coverage. This is a game-changer!

I've been advocating for this in the Discord community for a while now, and it's great to see it come to fruition. Congratulations to the Dagger team!

**Comment by XBeg9** [comment]
I am exploring the possibility of replacing Tilt's `live_update` feature with Dagger to eliminate the need for Dockerfiles. Specifically, I aim to establish a host-container synchronization mechanism that mounts a directory and monitors specific subdirectories for changes. For instance, when the `poetry.lock` file is modified, I would like to automatically execute commands such as `poetry install`.

In Tilt, this functionality is configured in the `Tiltfile` as follows:

```python
live_update=[
    sync('./api/', '/app/api/'),
    run(
        'pip install -r /app/requirements.txt',
        trigger=['./api/requirements.txt']
    )
]
```

**Comment by beingminimal** [comment]
@shykes @levlaz Any consideration or ETA for this feature? Awaiting your response. Thanks

**Comment by jbw-vtl** [comment]
Also quite keen to enable some form of mounting local files into a dagger service.

Specifically, we are looking to build a better experience for mk docs building, utilizing dagger to both build and deploy the static sites but also use it while developing the docs

**Comment by n2p5** [comment]
Posting this here for posterity (hopefully others find it useful)

I got a hack working last night, it _works_ but it requires the use of `--progress plain`

For my go project, I can use `entr` to reload `dagger call dev`

```find . -name "*.go" -type f | entr -r dagger --progress plain call dev --src=. up```

So I can hot-reload, and that part works great, and I get a basic terminal UI for it. Its so nice that my CI builds and my local dev builds share the same DNA now :) 

I'd love to help collaborate on something to make this dream a reality.

**Comment by aweris** [comment]
We're also interested in this feature. In our monorepo, we're unhappy with tilt and looking for alternative solutions. If we have this feature in dagger, that would be wonderful

**Comment by beingminimal** [comment]
> We're also interested in this feature. In our monorepo, we're unhappy with tilt and looking for alternative solutions. If we have this feature in dagger, that would be wonderful

Please check https://github.com/dagger/dagger/issues/8987

**Comment by gagemillerlob** [comment]
I am also interested in using dagger as a tilt alternative. I love tilt features but I don't like the confusing language and would like to use dagger instead.

Tilt features that would work great with dagger:
- hot reload on file watch
- file sync to live container
- deployments to k8s or docker compose with immutable tags
- local web ui for log view, build status and custom dynamic buttons to trigger dagger functions

I would love to replace all my Tiltfiles, makefiles and bash scripts with dagger.

**Comment by jedevc** [comment]
Heya y'all, I've written up a proposal for how we could build these features into dagger in https://github.com/dagger/dagger/issues/10895.

There's a lot to unpack, since there's a lot of use cases - but I'm looking for feedback that this is the right *shape* before starting to build on this, none of it is particularly trivial, and it'll be a pain to have to redo large chunks of it.

**Comment by shykes** [comment]
FYI we are reviving this with a new design: https://github.com/dagger/dagger/issues/13570
